flow/display_list_canvas_unittests.cc

```diff
@@ -1766,32 +1778,80 @@ class CanvasCompareTester {
       caseP.render_to(ref_canvas, testP);
       sk_sp<SkPicture> ref_x2_picture =
           sk_x2_recorder.finishRecordingAsPicture();
-      RenderSurface ref_x2_surface =
+      std::unique_ptr<RenderSurface> ref_x2_surface =
           env.MakeSurface(bg, TestWidth2, TestHeight2);
-      SkCanvas* ref_x2_canvas = ref_x2_surface.canvas();
+      SkCanvas* ref_x2_canvas = ref_x2_surface->canvas();
       ref_x2_canvas->scale(TestScale, TestScale);
       ref_x2_picture->playback(ref_x2_canvas);
-      const SkPixmap* ref_x2_pixels = ref_x2_surface.pixmap();
+      const SkPixmap* ref_x2_pixels = ref_x2_surface->pixmap();
       ASSERT_EQ(ref_x2_pixels->width(), TestWidth2) << info;
       ASSERT_EQ(ref_x2_pixels->height(), TestHeight2) << info;
       ASSERT_EQ(ref_x2_pixels->info().bytesPerPixel(), 4) << info;
 
       DisplayListBuilder builder_x2(TestBounds);
       caseP.render_to(builder_x2, testP);
       sk_sp<DisplayList> display_list_x2 = builder_x2.Build();
-      RenderSurface test_x2_surface =
+      std::unique_ptr<RenderSurface> test_x2_surface =
           env.MakeSurface(bg, TestWidth2, TestHeight2);
-      SkCanvas* test_x2_canvas = test_x2_surface.canvas();
+      SkCanvas* test_x2_canvas = test_x2_surface->canvas();
       test_x2_canvas->scale(TestScale, TestScale);
       display_list_x2->RenderTo(test_x2_canvas);
-      compareToReference(test_x2_surface.pixmap(), ref_x2_pixels,
+      compareToReference(test_x2_surface->pixmap(), ref_x2_pixels,
                          info + " (Both rendered scaled 2x)", nullptr, nullptr,
                          bg, TestWidth2, TestHeight2, false);
     }
   }
 
+  static void checkGroupOpacity(const RenderEnvironment& env,
+                                sk_sp<DisplayList> display_list,
+                                const SkPixmap* ref_pixmap,
+                                const std::string info,
+                                SkColor bg) {
+    SkScalar opacity = 128.0 / 255.0;
+
+    std::unique_ptr<RenderSurface> group_opacity_surface = env.MakeSurface(bg);
+    SkCanvas* group_opacity_canvas = group_opacity_surface->canvas();
+    display_list->RenderTo(group_opacity_canvas, opacity);
+    const SkPixmap* group_opacity_pixmap = group_opacity_surface->pixmap();
+
+    ASSERT_EQ(group_opacity_pixmap->width(), TestWidth) << info;
+    ASSERT_EQ(group_opacity_pixmap->height(), TestHeight) << info;
+    ASSERT_EQ(group_opacity_pixmap->info().bytesPerPixel(), 4) << info;
+
+    ASSERT_EQ(ref_pixmap->width(), TestWidth) << info;
+    ASSERT_EQ(ref_pixmap->height(), TestHeight) << info;
+    ASSERT_EQ(ref_pixmap->info().bytesPerPixel(), 4) << info;
+
+    int pixels_touched = 0;
+    int pixels_different = 0;
+    int fudge = env.info().bytesPerPixel() < 4 ? 9 : 2;
```

Comment:
Please add a comment explaining this.

Reply:
It's dead code actually. Oops.

Reply:
(Ignore previous comment about it being obsolete - there was a fudgeCompare function which I obsoleted, but this one remains.)